client/remote.go

```diff
@@ -16,17 +16,113 @@ import (
 	"github.com/miekg/dns"
 )
 
+// connPoolType is a async safe pool of established gokeyless Conn
+// so we don't need to do TLS handshake unnecessarily.
+type connPoolType struct {
+	pool map[string]*Conn
+	sync.RWMutex
+}
+
+// connPool keeps all active Conn
+var connPool connPoolType
+
 // A Remote represents some number of remote keyless server(s)
 type Remote interface {
-	Dial(*Client) (*gokeyless.Conn, error)
+	Dial(*Client) (*Conn, error)
 	Add(Remote) Remote
 }
 
+// A Conn represents a long-lived client connection to a keyserver.
+type Conn struct {
+	*gokeyless.Conn
+	addr string
+	ch   chan bool // a channel that will be read by health-check goroutine
+}
+
 // A singleRemote is an individual remote server
 type singleRemote struct {
-	net.Addr
-	ServerName string
-	conn       *gokeyless.Conn
+	net.Addr          // actual address
+	ServerName string // hostname for TLS verification
+}
+
+func init() {
+	connPool = connPoolType{
+		pool: make(map[string]*Conn),
+	}
+}
+
+// NewConn creates a new Conn based on a gokeyless.Conn
+func NewConn(addr string, conn *gokeyless.Conn) *Conn {
+	ch := make(chan bool)
+	c := Conn{
+		Conn: conn,
+		addr: addr,
+		ch:   ch,
+	}
+
+	go healthchecker(&c, ch, 1*time.Second)
+	ch <- true
+	return &c
+}
+
+// Close closes a Conn
+func (conn *Conn) Close() {
+	close(conn.ch)
+	conn.Conn.Close()
+	connPool.Remove(conn.addr)
+}
+
+// healthchecker is a recurrent timer function that tests the connections
+func healthchecker(conn *Conn, ch chan bool, after time.Duration) {
+	select {
+	case start, ok := <-ch:
+		if ok && start {
+			time.Sleep(after)
+			if !conn.Conn.IsClosed() {
+				err := conn.Ping(nil)
+				if err != nil {
+					log.Debug("health check ping failed:", err)
+					// shut down the conn
+					conn.Close()
+					return
+				}
+				log.Debug("start a new health check timer")
+				// start a new timer
+				go healthchecker(conn, ch, after)
+				ch <- true
+			}
+		}
+	}
+}
+
+// Get returns a Conn from the pool if there is any.
+func (p connPoolType) Get(key string) *Conn {
```

Comment:
I think you want these to be pointer receivers. Otherwise I think you're copying the mutex and doing all sorts of bad things.